packages/server/test/preHook.test.ts

```diff
@@ -0,0 +1,167 @@
+    trpc
+      .router()
+      .query('foo', {
+        resolve() {
+          return 'bar';
+        },
+      })
+      .preHook(preHook),
+  );
+
+  expect(await client.query('foo')).toBe('bar');
+  expect(preHook).toHaveBeenCalledTimes(0);
+  close();
+});
+
+test('allows you to throw an error (e.g. auth)', async () => {
+  type Context = {
+    user?: {
+      id: number;
+      name: string;
+      isAdmin: boolean;
+    };
+  };
+  const resolverMock = jest.fn();
+
+  const headers: Record<string, string | undefined> = {};
+
+  const { client, close } = routerToServerAndClient(
+    trpc
+      .router<Context>()
+      .query('foo', {
+        resolve() {
+          return 'bar';
+        },
+      })
+      .merge(
+        'admin.',
+        trpc
+          .router<Context>()
+          .preHook(({ ctx }) => {
+            if (!ctx.user?.isAdmin) {
+              throw httpError.unauthorized();
+            }
+          })
+          .query('secretPlace', {
+            resolve() {
+              resolverMock();
+
+              return 'a key';
+            },
+          }),
+      ),
+    {
+      createContext({ req }) {
+        if (req.headers.authorization === 'meow') {
+          return {
+            user: {
+              id: 1,
+              name: 'KATT',
+              isAdmin: true,
+            },
+          };
+        }
+        return {};
+      },
+      getHeaders: () => headers,
+    },
+  );
+
+  expect(await client.query('foo')).toBe('bar');
+  await expect(client.query('admin.secretPlace')).rejects.toMatchObject({
+    res: { status: 401 },
+  });
+  expect(resolverMock).toHaveBeenCalledTimes(0);
+  headers.authorization = 'meow';
+  expect(await client.query('admin.secretPlace')).toBe('a key');
+  expect(resolverMock).toHaveBeenCalledTimes(1);
+  close();
+});
+
+test('child routers + hook call order', async () => {
+  const preHookInParent = jest.fn();
+  const preHookInChild = jest.fn();
+  const preHookInGrandChild = jest.fn();
+  const { client, close } = routerToServerAndClient(
+    trpc
+      .router()
+      .preHook(preHookInParent)
+      .query('name', {
+        resolve() {
+          return 'GrandPa';
+        },
+      })
+      .merge(
+        'child.',
+        trpc
+          .router()
+          .preHook(preHookInChild)
+          .query('name', {
+            resolve() {
+              return 'Parent';
+            },
+          })
+          .merge(
+            'child.',
+            trpc
+              .router()
+              .preHook(preHookInGrandChild)
+              .query('name', {
+                resolve() {
+                  return 'GrandChild';
+                },
+              }),
+          ),
+      ),
+  );
+
+  expect(await client.query('child.child.name')).toBe('GrandChild');
+  expect(preHookInParent).toHaveBeenCalledTimes(1);
+  expect(preHookInChild).toHaveBeenCalledTimes(1);
+  expect(preHookInGrandChild).toHaveBeenCalledTimes(1);
```

Comment:
nice. so it will call each of those "middleware" hooks when going to a subrouter. That's perfect.